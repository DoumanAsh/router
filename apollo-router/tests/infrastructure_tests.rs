use apollo_router_core::{prelude::*, RouterBridgeQueryPlanner};
use std::sync::Arc;

#[test]
fn test_starstuff_supergraph_is_valid() {
    include_str!("../../examples/graphql/supergraph.graphql")
        .parse::<graphql::Schema>()
        .expect(
            r#"Couldn't parse the supergraph example.
This file is being used in the router documentation, as a quickstart example.
Make sure it is accessible, and the configuration is working with the router."#,
        );

    insta::assert_snapshot!(include_str!("../../examples/graphql/supergraph.graphql"));
}

#[tokio::test]
async fn test_query_planning_did_not_change() {
    let schema = Arc::new(
        include_str!("../../examples/graphql/supergraph.graphql")
            .parse::<graphql::Schema>()
            .unwrap(),
    );

    let query = r#"query TopProducts($first: Int) { topProducts(first: $first) { upc name reviews { id product { name } author { id name } } } }"#;

    let planner = RouterBridgeQueryPlanner::new(schema);
    let query_plan = planner
        .get(query.into(), None, Default::default())
        .await
        .unwrap();

    // Using println!() here so it shows up if the test fails (insta doesn't support messages in assert!)
    println!("/!\\ If this test fails, you want to double check the benchmark mocks, and run the benchmarks locally /!\\");
    insta::assert_debug_snapshot!(query_plan);
}
