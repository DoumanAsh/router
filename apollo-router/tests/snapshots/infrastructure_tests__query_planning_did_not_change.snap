---
source: apollo-router/tests/infrastructure_tests.rs
assertion_line: 35
expression: query_plan
---
QueryPlan {
    root: Sequence {
        nodes: [
            Fetch(
                FetchNode {
                    service_name: "products",
                    requires: [],
                    variable_usages: [
                        "first",
                    ],
                    operation: "query TopProducts__products__0($first:Int){topProducts(first:$first){__typename upc name}}",
                    operation_kind: Query,
                },
            ),
            Flatten(
                FlattenNode {
                    path: Path(
                        [
                            Key(
                                "topProducts",
                            ),
                            Flatten,
                        ],
                    ),
                    node: Fetch(
                        FetchNode {
                            service_name: "reviews",
                            requires: [
                                InlineFragment(
                                    InlineFragment {
                                        type_condition: Some(
                                            "Product",
                                        ),
                                        selections: [
                                            Field(
                                                Field {
                                                    alias: None,
                                                    name: "__typename",
                                                    selections: None,
                                                },
                                            ),
                                            Field(
                                                Field {
                                                    alias: None,
                                                    name: "upc",
                                                    selections: None,
                                                },
                                            ),
                                        ],
                                    },
                                ),
                            ],
                            variable_usages: [],
                            operation: "query TopProducts__reviews__1($representations:[_Any!]!){_entities(representations:$representations){...on Product{reviews{id product{__typename upc}author{__typename id}}}}}",
                            operation_kind: Query,
                        },
                    ),
                },
            ),
            Parallel {
                nodes: [
                    Flatten(
                        FlattenNode {
                            path: Path(
                                [
                                    Key(
                                        "topProducts",
                                    ),
                                    Flatten,
                                    Key(
                                        "reviews",
                                    ),
                                    Flatten,
                                    Key(
                                        "product",
                                    ),
                                ],
                            ),
                            node: Fetch(
                                FetchNode {
                                    service_name: "products",
                                    requires: [
                                        InlineFragment(
                                            InlineFragment {
                                                type_condition: Some(
                                                    "Product",
                                                ),
                                                selections: [
                                                    Field(
                                                        Field {
                                                            alias: None,
                                                            name: "__typename",
                                                            selections: None,
                                                        },
                                                    ),
                                                    Field(
                                                        Field {
                                                            alias: None,
                                                            name: "upc",
                                                            selections: None,
                                                        },
                                                    ),
                                                ],
                                            },
                                        ),
                                    ],
                                    variable_usages: [],
                                    operation: "query TopProducts__products__2($representations:[_Any!]!){_entities(representations:$representations){...on Product{name}}}",
                                    operation_kind: Query,
                                },
                            ),
                        },
                    ),
                    Flatten(
                        FlattenNode {
                            path: Path(
                                [
                                    Key(
                                        "topProducts",
                                    ),
                                    Flatten,
                                    Key(
                                        "reviews",
                                    ),
                                    Flatten,
                                    Key(
                                        "author",
                                    ),
                                ],
                            ),
                            node: Fetch(
                                FetchNode {
                                    service_name: "accounts",
                                    requires: [
                                        InlineFragment(
                                            InlineFragment {
                                                type_condition: Some(
                                                    "User",
                                                ),
                                                selections: [
                                                    Field(
                                                        Field {
                                                            alias: None,
                                                            name: "__typename",
                                                            selections: None,
                                                        },
                                                    ),
                                                    Field(
                                                        Field {
                                                            alias: None,
                                                            name: "id",
                                                            selections: None,
                                                        },
                                                    ),
                                                ],
                                            },
                                        ),
                                    ],
                                    variable_usages: [],
                                    operation: "query TopProducts__accounts__3($representations:[_Any!]!){_entities(representations:$representations){...on User{name}}}",
                                    operation_kind: Query,
                                },
                            ),
                        },
                    ),
                ],
            },
        ],
    },
}
